rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,370 @@
+```
+
+The `hoodie.table.index.function.list` indicates the Index functions that are ready for use at query planning and
+execution time.
+
+#### Index Function Creation
+
+To create an Index Function, we introduce two ways:
+
+1. New SQL statement:
+
+```
+CREATE INDEX FUNCTION [ index_function_name ] ON table_name [ USING index_type ] ( { column_name | expression } );
+```
+
+`table_name`: The name of the Hudi table for creating an Index Function.
+`index_function_name`: An optional name for the Index Function. If no name is specified, Hudi internally creates a new
+name.
+`column_name`: The column name without transformation for creating an Index Function.
+`expression`: The SQL expression on a data column for creating an Index Function.
+`index_type`: The type of index metadata to generate for the Index Function. By default, file-level column stats is
+used.
+
+Example:
+
+```
+CREATE INDEX datestr ON hudi_table USING col_stats (DATE_FORMAT(timestamp, '%Y-%m-%d'));
+CREATE INDEX ON hudi_table (city_id);
+CREATE INDEX hour_of_day ON hudi_table (DATE_FORMAT(timestamp, '%H'));
+```
+
+2. Write config in ingestion: A new write config is introduced `hoodie.write.index.function.list` to allow a user to add
+   and update Index Functions.
+
+Example:
+
+```
+hoodie.write.index.function.list=timestamp,datestr,col_stats,DATE_FORMAT(timestamp, '%Y-%m-%d');timestamp,,,DATE_FORMAT(timestamp, '%H')
+```
+
+#### Index in Metadata Table
+
+For each Index Function, Hudi needs to populate the index metadata to facilitate file pruning. We use the metadata table
+to store these index metadata. The index metadata of each individual Index Function is stored under a new partition with
+the naming of the Index Function name and the index type, e.g., `datestr_col_stats`. For the default index type of
+file-level column stats, the index format and logic of existing column stats index is reused. Once an Index Function is
+created and the corresponding index metadata is populated in metadata table, the index metadata is always consistent
+with the data table during write transactions, using the same multi-table transaction for updating metadata table.
+
+We plan to also integrate Index Function with Hudi's Asynchronous Indexing so that the index metadata for an Index
+Function can be populated in an async way.
+
+#### File Pruning
+
+When a query with the predicates related to an Index Functions is being planned, Hudi intercepts such predicate to carry
+out the file pruning
+
+1. Identify the relevant Index Function based on the predicate and add a new predicate based on the Index Function name
+   if needed. Take the Index Function created by
+   SQL `CREATE INDEX datestr ON hudi_table USING col_stats (DATE_FORMAT(timestamp, '%Y-%m-%d'));`:
+
+- If the original data column of an Index Function is used, transform the predicate to a new predicate based on the
+  Index Function name if possible, e.g., `timestamp between 1666767600 and 1666897200`
```

Comment:
If two index functions defined on the same column, we choose which one to use and based on what logic